silkworm/capi/silkworm.cpp

```diff
@@ -473,171 +477,267 @@ class BlockProvider {
+                }
+                if (signal_check(signal_check_time)) {
+                    return SILKWORM_TERMINATION_SIGNAL;
+                }
+
+                ++block_number;
+                prefetched_blocks.pop_front();
             }
 
+            auto last_block_number = block_number - 1;
+            log::Info{"[4/12 Execution] Flushing state",  // NOLINT(*-unused-raii)
+                      log_args_for_exec_flush(state_buffer, max_batch_size, last_block_number)};
+            state_buffer.write_state_to_db();
+            // Always save the Execution stage progess when state batch is flushed
+            db::stages::write_stage_progress(txn, db::stages::kExecutionKey, last_block_number);
+
             if (last_executed_block) {
-                *last_executed_block = block->header.number;
+                *last_executed_block = last_block_number;
             }
+        }
+        return SILKWORM_OK;
+    } catch (const mdbx::exception& e) {
+        if (mdbx_error_code) {
+            *mdbx_error_code = e.error().code();
+        }
+        return SILKWORM_MDBX_ERROR;
+    } catch (const DecodingError&) {
+        return SILKWORM_DECODING_ERROR;
+    } catch (const std::exception& e) {
+        SILK_ERROR << "exception: " << e.what();
+        return SILKWORM_INTERNAL_ERROR;
+    } catch (...) {
+        SILK_ERROR << "unknown exception";
+        return SILKWORM_UNKNOWN_ERROR;
+    }
+}
+
+SILKWORM_EXPORT
+int silkworm_execute_blocks_perpetual(SilkwormHandle handle, MDBX_env* mdbx_env, uint64_t chain_id,
+                                      uint64_t start_block, uint64_t max_block, uint64_t batch_size,
+                                      bool write_change_sets, bool write_receipts, bool write_call_traces,
+                                      uint64_t* last_executed_block, int* mdbx_error_code) SILKWORM_NOEXCEPT {
+    if (!handle) {
+        return SILKWORM_INVALID_HANDLE;
+    }
+    if (!mdbx_env) {
+        return SILKWORM_INVALID_MDBX_ENV;
+    }
+    if (start_block > max_block) {
+        return SILKWORM_INVALID_BLOCK_RANGE;
+    }
+    SignalHandlerGuard signal_guard;
+
+    try {
+        // Wrap MDBX env into an internal *unmanaged* env, i.e. MDBX env is only used but its lifecycle is untouched
+        db::EnvUnmanaged unmanaged_env{mdbx_env};
+        auto txn = db::RWTxnManaged{unmanaged_env};
+        const auto db_path{unmanaged_env.get_path()};
+
+        db::Buffer state_buffer{txn, /*prune_history_threshold=*/0};
+        BoundedBuffer<std::optional<Block>> block_buffer{kMaxBlockBufferSize};
+        BlockProvider block_provider{&block_buffer, unmanaged_env, start_block, max_block};
+        boost::strict_scoped_thread<boost::interrupt_and_join_if_joinable> block_provider_thread(block_provider);
+
+        const size_t max_batch_size{batch_size};
+        auto signal_check_time{std::chrono::steady_clock::now()};
+
+        std::optional<Block> block;
```

Comment:
You can move line 693 just before the `while` block at line 701 just for better readability.